File: Common/Freezable.vb
Public Class Freezable
    Inherits MVVM.BindableBase

    Public Sub Freeze()
        If Me.FreezeCalled Then
            Exit Sub
        End If
        Me.FreezeCalled = True
        Me.OnFreezing()
        Me._IsFreezed = True
        Me.OnFroze()
    End Sub

    Protected Overridable Sub OnFreezing()

    End Sub

    Protected Overridable Sub OnFroze()

    End Sub

    Protected Sub VerifyWrite()
        Verify.False(Me._IsFreezed, "Cannot change a freezed object.")
    End Sub

    Protected Sub VerifyFrozen()
        Verify.True(Me._IsFreezed, "The object has to be frozen to perform this operation.")
    End Sub

#Region "IsFreezed Property"
    Private _IsFreezed As Boolean

    Public ReadOnly Property IsFrozen As Boolean
        Get
            Return Me._IsFreezed
        End Get
    End Property
#End Region

    Private FreezeCalled As Boolean

End Class

File: Common/Freezable.cs
namespace Ks.Common
{
    public class Freezable : MVVM.BindableBase
    {
        public void Freeze()
        {
            if (this.FreezeCalled)
                return;
            this.FreezeCalled = true;

            this.OnFreezing();
            this._IsFrozen = true;
            this.OnFrozen();
        }

        protected virtual void OnFreezing()
        {
        }

        protected virtual void OnFrozen()
        {
        }

        protected void VerifyWrite()
        {
            Verify.False(this.IsFrozen, "Cannot change a frozen object.");
        }

        protected void VerifyFrozen()
        {
            Verify.True(this.IsFrozen, "The object has to be frozen to perform this operation.");
        }

        private bool _IsFrozen;

        public bool IsFrozen
        {
            get
            {
                return this._IsFrozen;
            }
        }

        private bool FreezeCalled;
    }
}
